rcl_action/src/rcl_action/action_client.c

```diff
@@ -0,0 +1,723 @@
+      if (RCL_RET_OK != ret) {
+        return RCL_RET_ERROR;
+      }
+    }
+    if (rcl_client_is_valid(&impl->result_client)) {
+      ret = rcl_client_fini(&impl->result_client, node);
+      if (ret != RCL_RET_OK) {
+        return RCL_RET_ERROR;
+      }
+    }
+    if (rcl_subscription_is_valid(&impl->feedback_subscription)) {
+      ret = rcl_subscription_fini(&impl->feedback_subscription, node);
+      if (ret != RCL_RET_OK) {
+        return RCL_RET_ERROR;
+      }
+    }
+    if (rcl_subscription_is_valid(&impl->status_subscription)) {
+      ret = rcl_subscription_fini(&impl->status_subscription, node);
+      if (ret != RCL_RET_OK) {
+        return RCL_RET_ERROR;
+      }
+    }
+    rcl_allocator_t * allocator = &impl->options.allocator;
+    allocator->deallocate(action_client->impl, allocator->state);
+    action_client->impl = NULL;
+  }
+  RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Action client finalized");
+  return RCL_RET_OK;
+}
+
+rcl_action_client_options_t
+rcl_action_client_get_default_options(void)
+{
+  static rcl_action_client_options_t default_options;
+  default_options.goal_service_qos = rmw_qos_profile_services_default;
+  default_options.cancel_service_qos = rmw_qos_profile_services_default;
+  default_options.result_service_qos = rmw_qos_profile_services_default;
+  default_options.feedback_topic_qos = rmw_qos_profile_default;
+  default_options.status_topic_qos = rcl_action_qos_profile_status_default;
+  default_options.allocator = rcl_get_default_allocator();
+  return default_options;
+}
+
+rcl_ret_t
+rcl_action_send_goal_request(
+  const rcl_action_client_t * action_client,
+  const void * ros_goal_request)
+{
+  RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Sending action goal request");
+  if (!rcl_action_client_is_valid(action_client)) {
+    return RCL_RET_ACTION_CLIENT_INVALID;
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(ros_goal_request, RCL_RET_INVALID_ARGUMENT);
+  int64_t ignored_sequence_number;
+  rcl_ret_t ret = rcl_send_request(
+    &action_client->impl->goal_client,
+    ros_goal_request, &ignored_sequence_number);
+  if (RCL_RET_OK != ret) {
+    return RCL_RET_ERROR;
+  }
+  RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Action goal request sent");
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_action_take_goal_response(
+  const rcl_action_client_t * action_client,
+  void * ros_goal_response)
+{
+  RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Taking action goal response");
+  if (!rcl_action_client_is_valid(action_client)) {
+    return RCL_RET_ACTION_CLIENT_INVALID;
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(ros_goal_response, RCL_RET_INVALID_ARGUMENT);
+  rmw_request_id_t ignored_request_header;
+  rcl_ret_t ret = rcl_take_response(
+    &action_client->impl->goal_client,
+    &ignored_request_header, ros_goal_response);
+  if (RCL_RET_OK != ret) {
+    if (RCL_RET_CLIENT_TAKE_FAILED == ret) {
+      ret = RCL_RET_ACTION_CLIENT_TAKE_FAILED;
+    } else {
+      ret = RCL_RET_ERROR;
```

Comment:
Same comment about returning original error code

Reply:
Same about documentation. I'm fine either way, your call.

Reply:
It can stay the way it is. It looks like existing code in RCL also returns `RCL_RET_ERROR` instead of the original code.

https://github.com/ros2/rcl/blob/4d8cb487f86bc709c6c0988cdc3ce80544db4b31/rcl/src/rcl/client.c#L86